### PR TITLE
Drop support for pre-8.0 GHCs

### LINKED ISSUE #110: Drop support for pre-8.0 versions of GHC in the `5` branch
This originally came out of a discussion in https://github.com/ekmett/semigroupoids/issues/130, which discusses some additional changes to `bifunctors` needed for the benefit of a future `foldable1-classes-compat` package. In particular, we wish to split out the `Data.Bifunctor`, `Data.Bifoldable`, and `Data.Bitraversable` modules from `bifunctors` into a more minimal `bifunctor-classes-compat` package with fewer dependencies, and then have `bifunctors` re-export these modules. There is one catch: `cabal`'s `reexported-modules` feature only supports GHC 7.10 and later. As a result, we have decided that this is as good of a time as any to drop pre-8.0 support in `bifunctors`, which is perhaps overdue as it is.

Note that this really only matters for the [`5` branch](https://github.com/ekmett/bifunctors/tree/5) of `bifunctors`. The `main` branch only support GHC 8.6 and later, so none of these changes would apply to that branch.

### PR DESCRIPTION
**Pull request by RyanGlScott**
Fixes #110.

### CODE REVIEW

**Comment by phadej** [comment]
I tried this branch, and it said `273f89500762239c6ebaa7cd8f56ba300b24aaed` is not in tree (of `bifunctor-classes-compat`).

File: bifunctors.cabal [hunk 1]
```diff
@@ -54,42 +39,27 @@ flag tagged
...
   if impl(ghc<8.1)
     hs-source-dirs: old-src/ghc801
     exposed-modules:
       Data.Bifoldable
       Data.Bitraversable
 
-  if impl(ghc>=7.2) && impl(ghc<7.5)
-    build-depends: ghc-prim == 0.2.0.0
-
   exposed-modules:
```
Comment (phadej) [inline bifunctors.cabal:63]: The ` Paths_bifunctors` could be removed from `other-modules`, can't it?

File: bifunctors.cabal [hunk 2]
```diff
@@ -54,42 +39,27 @@ flag tagged
...
     comonad             >= 5.0.7 && < 6,
     containers          >= 0.2   && < 0.7,
-    template-haskell    >= 2.4   && < 2.20,
+    template-haskell    >= 2.11  && < 2.20,
     th-abstraction      >= 0.4.2.0 && < 0.5,
-    transformers        >= 0.3   && < 0.7
+    transformers        >= 0.5   && < 0.7
 
   if !impl(ghc > 8.2)
     build-depends: transformers-compat >= 0.5 && < 0.8
```
Comment (phadej) [inline bifunctors.cabal:52]: This is there to provide orphan instances through `Data.Bifoldable` and `Data.Bitraversable`.

However looking at https://hackage.haskell.org/package/transformers-compat-0.7.2/docs/src/Control.Monad.Trans.Instances.html and https://hackage.haskell.org/package/transformers-compat-0.7.2/changelog it looks the lower bound needs to be 0.6?

> Backport changes from transformers-0.5.4 (i.e., add Bifoldable and Bitraversable instances for Data.Functor.Constant)

other instances in `Control.Monad.Trans.Instances` don't seem to be relevant to `Bifunctor` family of classes

File: bifunctors.cabal [hunk 3]
```diff
@@ -54,42 +39,27 @@ flag tagged
 
 library
   hs-source-dirs: src
-  include-dirs: include
-  includes: bifunctors-common.h
   build-depends:
-    base                >= 4.3   && < 5,
+    base                >= 4.9   && < 5,
     base-orphans        >= 0.8.4 && < 1,
```
Comment (phadej) [inline bifunctors.cabal:44]: This is used only in `Data.Bitraversable`. Would it make sense to move into `!impl(ghc > 8.2)` conditional?

File: bifunctors.cabal [hunk 4]
```diff
@@ -54,42 +39,27 @@ flag tagged
...
     comonad             >= 5.0.7 && < 6,
     containers          >= 0.2   && < 0.7,
```
Comment (phadej) [inline bifunctors.cabal:46]: This lower bound could be 0.5.7.1 (which is bundled with GHC-8.0.2). There doesn't seem to be install plans with older containers anyway.

File: bifunctors.cabal (hunk 3 above)
Comment (RyanGlScott) [reply bifunctors.cabal]: Yes, I agree. Done.

File: bifunctors.cabal (hunk 4 above)
Comment (RyanGlScott) [reply bifunctors.cabal]: I've bumped the lower bound on `containers` to `0.5.7.1`.

File: bifunctors.cabal (hunk 1 above)
Comment (RyanGlScott) [reply bifunctors.cabal:63]: Yes, you are quite right. I've removed `Paths_bifunctors` from this list.

File: bifunctors.cabal (hunk 2 above)
Comment (RyanGlScott) [reply bifunctors.cabal]: I've bumped the lower version bounds on `transformers-compat` to `0.6`.

**Pull request merged**